Grafana operator dashboard

subkanthi opened the pull request:
- Added Grafana operator.
- Added Prometheus operator(workaround bug for CRD too long error), Added prometheus CRD, RBAC, operator and Prometheus app as separate argoCD apps.
- Added Grafana datasource and dashboards.

The prometheus pod gets to this state
```
   shard 0: pod prometheus-prometheus-0: containers with incomplete status: [init-config-reloader]
```
and `argocd app sync prometheus` is required one more time to get it to an healthy state.

```[ClickHouse in Kubernetes Queries]``` - Dashboard does not load.

ToDo - 

- check for Duplicate `ClickHouseInstallation`
- Add infra folder(move Grafana and Prometheus ) , leave clickhouse-operator and clickhouse outside.
- check why prometheus needs to be synced.
-

subkanthi committed:
Updated grafana operator

subkanthi committed:
Added grafana dashboard and data source

subkanthi committed:
Changed grafana dashboard apiVersion to v1beta1

subkanthi committed:
Added grafana dashboards from clickhouse operator.

subkanthi committed:
Fixed grafana datasource for grafana operator.

subkanthi committed:
Fixed grafana datasource for grafana operator.

subkanthi committed:
Fixed crd name in grafana datasource.

subkanthi committed:
Added spec.instanceSelector to all grafana.

subkanthi committed:
Added service to clickhouseInstallation so that the datasource can be accessed.

subkanthi committed:
Removed type Loadbalancer for CH service.

subkanthi committed:
Removed type Loadbalancer for CH service.

subkanthi committed:
Updated clickhouse service endpoint for grafana datasource.

subkanthi committed:
Updated grafana username/password to non admin user.

subkanthi committed:
Update clickhouse service endpoint.

subkanthi committed:
Update clickhouse service endpoint.

subkanthi committed:
Removed array as datasource.

subkanthi committed:
Added authentication to grafana datasource.

subkanthi committed:
set argocd sync waves to different apps.

subkanthi committed:
Updated grafana datasource URL and username/password.

subkanthi committed:
Added prometheus target and scrape config.

subkanthi committed:
Set sync-wave priority to -2 for prometheus operator.

subkanthi committed:
Separate out prometheus operator and prometheus manifest files.

subkanthi committed:
Add prometheus operator CRD helm chart.

subkanthi committed:
Add prometheus operator CRD helm chart.

subkanthi committed:
Fixed create-stack.sh

subkanthi committed:
Set synch wave priority for prometheus.

subkanthi committed:
Add serverSideApply flag for prometheus operator CRD to workaround the crd too long bug.

subkanthi committed:
Add argocd annotation to replace long CRD name.

subkanthi committed:
fixed namespace in prometheus operator crd.

subkanthi committed:
Added kustomization.yaml to install prometheus CRDs

subkanthi committed:
Removed prometheus helm chart.

subkanthi committed:
Added gitignore

subkanthi committed:
Added stripped down CRD for prometheus operator

subkanthi committed:
Removed kustomization.yaml for prometheus operator

subkanthi committed:
enable helm chart and move stripped down crd to templates folder.

subkanthi committed:
Added documentation.

subkanthi committed:
Add prometheus operator argocd application manifest.

subkanthi committed:
Add bundle.yaml to install prometheus operator.

subkanthi committed:
Remove CRD definition from prometheus operator manifest.

subkanthi committed:
Added Prometheus operator RBAC settings.

subkanthi committed:
Set namespace for prometheus Role binding.

subkanthi committed:
Set namespace for prometheus Role and service account.

subkanthi committed:
Set namespace for prometheus Role and service account.

subkanthi committed:
Set Prometheus datasource in Grafana to prometheus:9090

subkanthi committed:
Updated README and delete-stack-apps.sh to include prometheus-rbac.

subkanthi committed:
Merged changes from main

subkanthi committed:
Added instructions for grafana port-forwarding and opening UI.

subkanthi committed:
Added script to install Argo CD.

subkanthi committed:
Updated instructions in README.md

hodgesrm commented:
Hi Kanthi, an immediate comment. Can you squash your commits into a single commit so that the log is cleaner when you merge? I'll start reviewing the code meanwhile.

hodgesrm commented on apps/clickhouse/demo.yaml:
```diff
@@ -1,9 +1,39 @@
+## Prometheus operator podMonitor
```
We need to figure out how to add this in a way that works when prometheus is not installed. It should be possible to bring up clickhouse-operator + zookeeper + clickhouse, for example.

subkanthi replied:
The `PodMonitor` can be separated out into a separate argocd app or into the other ones.
But the metrics has to be enabled in ClickHouseInstallation
```
      - name: server
        spec:
          containers:
          - name: clickhouse
            image: altinity/clickhouse-server:22.3.15.34.altinitystable
            ports:
              - name: "metrics"
                containerPort: 9363
```

hodgesrm commented on apps/prometheus-rbac/deployment.yaml:
```diff
@@ -0,0 +1,54 @@
+subjects:
+  - kind: ServiceAccount
+    name: prometheus
+    namespace: ch
```
Namespaces should not be hard-coded. It should be possible to pick the namespace.

subkanthi replied:
This is the error on the `prometheus-operator`, it expects the namespace
```
rbac.authorization.k8s.io  ClusterRoleBinding  ch         prometheus-operator  Failed     SyncFailed        ClusterRoleBinding.rbac.authorization.k8s.io "prometheus-operator" is invalid: subjects[0].namespace: Required value
```

subkanthi replied:
argocd suggested way is to define in the application manifest `spec.destination.namespace`
```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: foo
  namespace: argocd
spec:
  destination:
    namespace: bar
  syncPolicy:
    syncOptions:
      createNamespace:
        enabled: true
        labels:
          io.foo.bar/this: labelValue
          io.foo.bar/that: labelValue
      validate:
        enabled: false
```

But still trying to figure out a way to do it from the argocd cli.

hodgesrm commented on apps/grafana-operator/README.md:
````diff
@@ -1,32 +1,60 @@
+```shell
+helm upgrade -i grafana-operator oci://ghcr.io/grafana-operator/helm-charts/grafana-operator --version v5.3.0
 ```
-argocd app create grafana-operator \
````
These instructions should be retained.

subkanthi replied:
added back.

hodgesrm commented on apps/grafana-operator/Chart.yaml:
```diff
@@ -1,5 +1,5 @@
 apiVersion: v2
-name: prometheus
```
It looks as if this chart copies the YAML fully from the original Helm chart. It should should be possible to reference it if it's in a public help repo. We don't want to maintain detailed files if we can avoid it.

subkanthi replied:
Maybe its my kind cluster setup, argocd did not install the kustomize yaml, so switched over to copying it .

hodgesrm commented on create-stack.sh:
```diff
@@ -4,33 +4,76 @@ echo "This command will install the analytic stack in namespace ${NS}"
 echo -n "Press enter or ^c to cancel: "
 read
 set -x
-argocd app create clickhouse-operator \
```
Rather than override this we should provide installation alternatives. The advantage of using argocd commands is that users can just refer to the cloud and load selected parts of the stack quickly.

subkanthi replied:
it still uses argocd commands in a for loop
```
declare -a applications=(prometheus-operator-crds prometheus-rbac prometheus-operator clickhouse-operator prometheus grafana-operator zookeeper clickhouse grafana grafana-datasource)

#helm upgrade -i grafana-operator oci://ghcr.io/grafana-operator/helm-charts/grafana-operator --version v5.3.0 --values apps/grafana-operator/values.yaml --namespace ${NS} --create-namespace
#sleep 5

for apps in "${applications[@]}"; do
  echo "*********** Creating ${apps} ***********"
  argocd app create ${apps} \
    --repo https://github.com/Altinity/argocd-examples-clickhouse.git --path apps/${apps} \
    --dest-server https://kubernetes.default.svc --dest-namespace ${NS} --revision grafana_operator_dashboard
done
```

hodgesrm commented on grafana_crd.yaml:
```diff
@@ -0,0 +1,5727 @@
+apiVersion: apiextensions.k8s.io/v1
```
Looks like cruft?

subkanthi replied:
This is part of the Grafana operator CRD, no changes were made.

subkanthi replied:
you are correct, removed.

hodgesrm reviewed (changes requested):
Thanks for the opportunity to review. Here are some general comments. 
1.) We should support installation of apps using 'argocd create' or using Application manifests & kubectl. 
2. Each app at the level of the apps directory should be independently installable. So it should be possible to pick, say clickhouse-operator & operator apps and bring up just those two. It allows user users to substitute different services including their own. 
3. Users should be able to install the app in any namespace. 
4. We should load helm charts by reference wherever possible to keep this project as simple as possible. To ensure this works we can pin versions.

subkanthi committed:
Removed namespace references in grafana and clickhouse

subkanthi committed:
reverted back namespace in prometheus operator

subkanthi committed:
Removed creating prometheus namespace.

subkanthi committed:
Reverted namespace back to clickhouse pod metrics and grafana.

subkanthi committed:
Removed grafana_crd.yaml

subkanthi committed:
Removed namespace for prometheus operator.

subkanthi committed:
Removed namespace for prometheus RBAC.

subkanthi committed:
Removed namespace for prometheus RBAC.

subkanthi committed:
reverted back namespace settings in prometheus-rbac.

subkanthi committed:
Updated sync-apps script.

subkanthi committed:
Add missing namespace for prometeheus-operator

subkanthi committed:
updated instructions for create-stack.sh with revision.

subkanthi committed:
Moved dependencies to infra folder.

subkanthi committed:
Moved dependencies to infra folder.

subkanthi committed:
Set sync waves to apps.

subkanthi committed:
Add prometheus operator to app of apps.

subkanthi committed:
ADd grafana apps to app of apps.

subkanthi committed:
Add namespace for operator crd.

subkanthi committed:
Set namespace for app of apps.

subkanthi committed:
Set targetRevision in apps.

subkanthi committed:
Remove name in spec/destination for apps.

subkanthi committed:
Set cluster name for crds

subkanthi committed:
Fix path name

subkanthi committed:
Fix path name

subkanthi committed:
Add another application yaml file inside prometheus-operator-crds

subkanthi committed:
Set path for prometheus operator CRD child app.

subkanthi committed:
Set path for prometheus operator CRD child app.

subkanthi committed:
Add script to create infrastructure components(grafana, grafana operator, prometheus CRD, prometheus operator)

subkanthi committed:
Add script to create infrastructure components(grafana, grafana operator, prometheus CRD, prometheus operator)

subkanthi committed:
Added grafana and grafana-datasource to create-infra-stack.sh

subkanthi committed:
Added script to delete infra apps.

subkanthi committed:
Removed clickhouse queries dashboard

hodgesrm committed:
Update infra stack create and delete scripts

Changed the scripts to drop the directory name and extended the
options so we can generalize this style of script across multiple
services.

hodgesrm committed:
Add simple helm chart to install ClickHouse
